lib/collection/src/shard/remote_shard.rs

```diff
@@ -27,7 +68,162 @@ impl ShardOperation for &RemoteShard {
         operation: CollectionUpdateOperations,
         wait: bool,
     ) -> CollectionResult<UpdateResult> {
-        todo!()
+        let mut client = self.points_client().await?;
+
+        let response = match operation {
+            CollectionUpdateOperations::PointOperation(point_ops) => match point_ops {
+                PointOperations::UpsertPoints(upsert_points) => {
```

Comment:
Same concern here - you might want to move inner match branches into appropriate functions or trait impls

Reply:
struggling a bit here to separate cleanly the conversion due to the side effects, I will try harder

Reply:
You can do it without side effects I think:
```rust
impl UpsertPointsInternal {
// .. some other methods here

pub fn from(upsert_points: UpsertPoints, shard: &RemoteShard) -> CollectionResult<Self> {
Ok(Self {
                        shard_id: shard.id,
                        upsert_points: Some(UpsertPoints {
                            collection_name: shard.collection_id.clone(),
                            wait: Some(true),
                            points: match upsert_points {
                                PointInsertOperations::PointsBatch(batch) => {
                                    return Err(CollectionError::service_error(
                                        "Operation not supported".to_string(),
                                    ))
                                }
                                PointInsertOperations::PointsList(list) => list
                                    .points
                                    .into_iter()
                                    .map(|id| id.try_into())
                                    .collect::<Result<Vec<_>, Status>>()?,
                            },
                        }),
                    })
}
}
```

Reply:
Then:
```rust
let request = tonic::Request::new(UpsertPointsInternal::from(upsert_points, &self)?;
client.upsert(request).await?
```

Reply:
Thanks for your input!

I ended up extracting the functions this way https://github.com/qdrant/qdrant/pull/505/commits/fc9d64b54a6ad4074e51888571847387444c9852

It was easiest way to have access to all the necessary type. I could not `impl UpsertPointsInternal` from `collections` as the type lives in `api`.

Reply:
Yes, sure, I noticed it only later